source/common/router/upstream_request.h

```diff
@@ -141,6 +144,9 @@ class UpstreamRequest : public Logger::Loggable<Logger::Id::router>,
   StreamInfo::StreamInfoImpl stream_info_;
   StreamInfo::UpstreamTiming upstream_timing_;
   const MonotonicTime start_time_;
+  // This is wrapped in an optional, since we want to avoid computing zero size headers when in
+  // reality we just didn't get a response back.
+  absl::optional<uint64_t> response_headers_size_{};
```

Comment:
optionally: what do you think of just initializing to zero?  I don't think you can have valid 0 length headers so not recording when 0 seems fine to me.
That said if we do optional lengths for bodies (where we may want to disambiguate between 0 and not-present) and you want this optional for consistency that's fine.

Reply:
See below, leaning toward avoiding detecting if body should be there and accounting zero sized bodies (IMO they don't add signal anyway).